File: apps/api/app/services/fetchers/tiktok.py
"""TikTok fetcher — video metrics via Content Posting API / Research API."""

import logging

import httpx

from app.services.fetchers.base import BaseFetcher, NormalizedMetrics

logger = logging.getLogger(__name__)

TIKTOK_API = "https://open.tiktokapis.com/v2"


class TikTokFetcher(BaseFetcher):
    """Fetch TikTok video metrics via the query endpoint.

    Uses /video/query/ with fields for view, like, comment, share counts.
    Requires OAuth 2.0 token with video.list scope.
    """

    platform_id = "tiktok"

    async def fetch(self, token: str, post_id: str) -> NormalizedMetrics | None:
        try:
            async with httpx.AsyncClient(timeout=15) as client:
                resp = await client.post(
                    f"{TIKTOK_API}/video/query/",
                    json={"filters": {"video_ids": [post_id]}},
                    params={
                        "fields": "view_count,like_count,comment_count,share_count,favorite_count",
                    },
                    headers={
                        "Authorization": f"Bearer {token}",
                        "Content-Type": "application/json",
                    },
                )

            if resp.status_code != 200:
                logger.warning("TikTok analytics error %d for video %s", resp.status_code, post_id)
                return None

            videos = resp.json().get("data", {}).get("videos", [])
            if not videos:
                logger.warning("TikTok video %s not found in response", post_id)
                return None

            video = videos[0]
            views = video.get("view_count", 0)
            likes = video.get("like_count", 0)
            comments = video.get("comment_count", 0)
            shares = video.get("share_count", 0)
            favorites = video.get("favorite_count", 0)

            engagements = likes + comments + shares + favorites
            engagement_rate = round(engagements / max(views, 1), 6)

            return NormalizedMetrics(
                impressions=views,
                engagements=engagements,
                engagement_rate=engagement_rate,
                saves_bookmarks=favorites,
                shares_reposts=shares,
                comments=comments,
                clicks=0,  # Not available in basic video query
                follows_gained=0,
                platform_specific={
                    "views": views,
                    "likes": likes,
                    "comments": comments,
                    "shares": shares,
                    "favorites": favorites,
                },
            )

        except Exception as exc:
            logger.warning("Failed to fetch TikTok metrics for video %s: %s", post_id, exc)
            return None

File: functions/src/lib/analytics/fetchers/tiktok.ts
/**
 * TikTok metrics fetcher — Content Posting API /video/query.
 */

import type { MetricsFetcher, FetchMetricsOptions, PlatformMetrics } from "./base.js";
import { handleApiResponse } from "./base.js";

interface TikTokVideoData {
  like_count?: number;
  comment_count?: number;
  share_count?: number;
  view_count?: number;
}

export const tiktokFetcher: MetricsFetcher = {
  platformId: "tiktok",

  async fetchMetrics(options: FetchMetricsOptions): Promise<PlatformMetrics> {
    const { platformPostId, accessToken } = options;

    const resp = await fetch(
      "https://open.tiktokapis.com/v2/video/query/",
      {
        method: "POST",
        headers: {
          Authorization: `Bearer ${accessToken}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          filters: { video_ids: [platformPostId] },
          fields: ["like_count", "comment_count", "share_count", "view_count"],
        }),
      }
    );

    handleApiResponse("tiktok", resp, platformPostId);

    const json = await resp.json() as {
      data?: { videos?: TikTokVideoData[] };
    };

    const video = json.data?.videos?.[0];
    if (!video) {
      return { impressions: 0, engagements: 0, saves: 0, shares: 0, clicks: 0, follows: 0, comments: 0 };
    }

    const views = video.view_count ?? 0;
    const likes = video.like_count ?? 0;
    const comments = video.comment_count ?? 0;
    const shares = video.share_count ?? 0;

    return {
      impressions: views,
      engagements: likes + comments + shares,
      saves: 0,
      shares,
      clicks: 0,
      follows: 0,
      comments,
    };
  },
};
